packages/next/src/client/form.tsx

```diff
@@ -0,0 +1,326 @@
+'use client'
+
+import { useEffect, type HTMLProps, type FormEvent } from 'react'
+import { useRouter } from './components/navigation'
+import { addBasePath } from './add-base-path'
+import { useIntersection } from './use-intersection'
+import { useMergedRef } from './use-merged-ref'
+import type { AppRouterInstance } from '../shared/lib/app-router-context.shared-runtime'
+
+const DISALLOWED_FORM_PROPS = ['method', 'encType', 'target'] as const
+
+type HTMLFormProps = HTMLProps<HTMLFormElement>
+type DisallowedFormProps = (typeof DISALLOWED_FORM_PROPS)[number]
+
+export type FormProps = {
+  /**
+   * `action` can be either a `string` or a function.
+   * - If `action` is a string, it will be interpreted as a path or URL to navigate to when the form is submitted.
+   *   The path will be prefetched when the form becomes visible.
+   * - If `action` is a function, it will be called when the form is submitted. See the [React docs](https://react.dev/reference/react-dom/components/form#props) for more.
+   */
+  action: NonNullable<HTMLFormProps['action']>
+  /**
+   * Whether submitting the form should replace the current `history` state instead of adding a new url into the stack.
+   * Only valid if `action` is a string.
+   *
+   * @defaultValue `false`
+   */
+  replace?: boolean
+  /**
+   * Override the default scroll behavior when navigating.
+   * Only valid if `action` is a string.
+   *
+   * @defaultValue `true`
+   */
+  scroll?: boolean
+} & Omit<HTMLFormProps, 'action' | DisallowedFormProps>
+
+export default function Form({
+  replace,
+  scroll,
+  ref: externalRef,
+  ...props
+}: FormProps) {
+  const actionProp = props.action
+
+  for (const key of DISALLOWED_FORM_PROPS) {
+    if (key in props) {
+      if (process.env.NODE_ENV === 'development') {
+        console.error(
+          `<Form> does not support changing \`${key}\`. ` +
+            (typeof actionProp === 'string'
+              ? `If you'd like to use it to perform a mutation, consider making \`action\` a function instead.\n` +
+                `Learn more: https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations`
+              : '')
+        )
+      }
+      delete (props as Record<string, unknown>)[key]
+    }
+  }
+
+  const router = useRouter()
+
+  const [setIntersectionRef, isVisible] = useIntersection({
+    rootMargin: '200px',
+    disabled: typeof actionProp !== 'string', // if we don't have an action path, we can't preload anything anyway.
```

Comment:
nit: maybe worth hoisting this assignment to something with a good name that we can leverage anywhere we disable behavior because it's not a "valid" action type for `next/form`. eg. `isGetForm` or something

Reply:
i was also thinking about that, but unfortunately i think typescript won't be smart enough to correlate the variable with the type, so we'll still need to check it everywhere...